Make KEP-4193 documentation updates (#45292)
kubernetes/website: content/en/docs/reference/access-authn-authz/service-accounts-admin.md

  without many constraints and have namespaced names, such configuration is
  usually portable.

## Bound service account tokens

ServiceAccount tokens can be bound to API objects that exist in the kube-apiserver.
This can be used to tie the validity of a token to the existence of another API object.
Supported object types are as follows:

* Pod (used for projected volume mounts, see below)
* Secret (can be used to allow revoking a token by deleting the Secret)

> * Node (beta in v1.30)

Author: One finicky part here - authenticating with node bound tokens is beta in 1.30, but actually issuing node bound tokens doesn't go beta until 1.31..

> * Node (beta in v1.30)

Reviewer: ```suggestion
* Node (in v1.30, creating new node-bound tokens is alpha, using existing node-bound tokens is beta)
```

  without many constraints and have namespaced names, such configuration is
  usually portable.

## Bound service account tokens

ServiceAccount tokens can be bound to API objects that exist in the kube-apiserver.
This can be used to tie the validity of a token to the existence of another API object.
Supported object types are as follows:

* Pod (used for projected volume mounts, see below)

> * Secret (used when generating legacy service account tokens)

Reviewer: ```suggestion
* Secret (can be used to allow revoking a token by deleting the Secret)
```

* Node (beta in v1.30)

> When a token is bound to an object, the objects `metadata.name` and `metadata.uid` are

Reviewer: ```suggestion
When a token is bound to an object, the object's `metadata.name` and `metadata.uid` are
```

stored as extra 'private claims' in the issued JWT.

When a bound token is presented to the kube-apiserver, the service account authenticator

> will extra and verify these claims.

Reviewer: ```suggestion
will extract and verify these claims.
```

If the referenced object no longer exists (or its `metadata.uid` does not match),
the request will not be authenticated.

### Additional metadata in Pod bound tokens

{{< feature-state feature_gate_name="ServiceAccountTokenPodNodeInfo" >}}

When a service account token is bound to a Pod object, additional metadata is also

> embedded into the token that indicates the name of the Node that the Pod is scheduled to.

Reviewer: ```suggestion
embedded into the token that indicates the value of the bound pod's `spec.nodeName` field,
and the uid of that Node, if available.
```

> This node information is **not** verified by the kube-apiserver, and is included to
> avoid integrators needing having to retrieve a full `Pod` object from the apiserver
> when inspecting a JWT.

Reviewer: ```suggestion
This node information is **not** verified by the kube-apiserver when the token is used for authentication.
It is included so integrators do not have to fetch Pod or Node API objects to check the associated Node name
and uid when inspecting a JWT.
```

### Verifying and inspecting private claims

The `TokenReview` API can be used to verify and extract private claims from a token:

1. First, assume you have a pod named `test-pod` and a service account named `my-sa`.
2. Create a token that is bound to this Pod:

```shell
kubectl create token my-sa --bound-object-kind="Pod" --bound-object-name="test-pod"
```

3. Copy this token into a new file named `tokenreview.yaml`:

```yaml
apiVersion: authentication.k8s.io/v1
kind: TokenReview
spec:
  token: <token from step 2>
```

4. Submit this resource to the apiserver for review:

```shell
kubectl create -o yaml -f tokenreview.yaml # we use '-o yaml' so we can inspect the output
```

You should see an output like below:

```yaml
apiVersion: authentication.k8s.io/v1
kind: TokenReview
metadata:
  creationTimestamp: null
spec:
  token: <token>
status:
  audiences:
  - https://kubernetes.default.svc.cluster.local
  authenticated: true
  user:

>     extra:

Reviewer: wouldn't the node name / uid and jti appear here?

Author: If all of the relevant feature gates are enabled, yes. I guess as these are all enabled by default now though, it does make sense to include them here.

  without many constraints and have namespaced names, such configuration is
  usually portable.

## Bound service account tokens

ServiceAccount tokens can be bound to API objects that exist in the kube-apiserver.
This can be used to tie the validity of a token to the existence of another API object.
Supported object types are as follows:

* Pod (used for projected volume mounts, see below)
* Secret (can be used to allow revoking a token by deleting the Secret)
* Node (beta in v1.30)

When a token is bound to an object, the object's `metadata.name` and `metadata.uid` are
stored as extra 'private claims' in the issued JWT.

When a bound token is presented to the kube-apiserver, the service account authenticator
will extract and verify these claims.
If the referenced object no longer exists (or its `metadata.uid` does not match),
the request will not be authenticated.

### Additional metadata in Pod bound tokens

> {{< feature-state feature_gate_name="ServiceAccountTokenPodNodeInfo" >}}

Reviewer: does this automatically pull the beta status, or do we have to indicate that? all the other examples I see have to put the version and state in: https://grep.app/search?q=%7B%7B%3C%20feature-state&filter[repo][0]=kubernetes/website&filter[path][0]=content/en/

Reviewer 2: The current proposed change looks good. It leverages the new `feature_gate_name` option which automates the feature status retrieval based on the information in the feature description file i.e. '[service-account-token-pod-node-info.md](https://github.com/kubernetes/website/blob/main/content/en/docs/reference/command-line-tools-reference/feature-gates/service-account-token-pod-node-info.md?plain=1)'.

Reviewer 3: (see https://kubernetes.io/docs/contribute/new-content/new-features/)
